bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java

```diff
@@ -0,0 +1,265 @@
+        logger.debug("Activating FMI Weather discovery service");
+        super.activate(configProperties);
+    }
+
+    @Override
+    @Modified
+    protected void modified(@Nullable Map<String, @Nullable Object> configProperties) {
+        super.modified(configProperties);
+    }
+
+    @Override
+    @Deactivate
+    public void deactivate() {
+        logger.debug("Deactivating FMI Weather discovery service");
+        super.deactivate();
+    }
+
+    @Override
+    protected void startScan() {
+        PointType location = null;
+        logger.debug("Starting FMI Weather discovery scan");
+        LocationProvider locationProvider = getLocationProvider();
+        location = locationProvider.getLocation();
+        if (location == null) {
+            logger.debug("LocationProvider.getLocation() is not set -> Will discover all stations");
+        }
+        createResults(location);
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        if (discoveryJob == null) {
+            discoveryJob = scheduler.scheduleWithFixedDelay(() -> {
+                PointType currentLocation = locationProvider.getLocation();
+                if (!Objects.equals(currentLocation, previousLocation)) {
+                    logger.debug("Location has been changed from {} to {}: Creating new discovery results",
+                            previousLocation, currentLocation);
+                    createResults(currentLocation);
+                    previousLocation = currentLocation;
+                }
+            }, 0, LOCATION_CHANGED_CHECK_INTERVAL_SECONDS, TimeUnit.SECONDS);
+            logger.debug("Scheduled FMI Weather location-changed discovery job every {} seconds",
+                    LOCATION_CHANGED_CHECK_INTERVAL_SECONDS);
+        }
+    }
+
+    public void createResults(@Nullable PointType location) {
+        createForecastForCurrentLocation(location);
+        createForecastsForCities(location);
+        createObservationsForStations(location);
+    }
+
+    private void createForecastForCurrentLocation(@Nullable PointType currentLocation) {
+        if (currentLocation != null) {
+            Map<String, Object> properties = new HashMap<>();
+            properties.put(LOCATION,
+                    String.format("%s,%s", currentLocation.getLatitude(), currentLocation.getLongitude()));
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(UID_LOCAL_FORECAST)
+                    .withLabel(String.format("FMI local weather forecast")).withProperties(properties).build();
+            thingDiscovered(discoveryResult);
+        }
+    }
+
+    private void createForecastsForCities(@Nullable PointType currentLocation) {
+        List<Location> filteredCities = new LinkedList<>();
+        CITIES_OF_FINLAND.stream().filter(location2 -> isClose(currentLocation, location2)).peek(station -> {
+            if (logger.isDebugEnabled()) {
```

Comment:
Are these log messages helpful for any user? Is the result not the same as displayed in PaperUI?

Reply:
True! fd518b4e9271b71fe63cc07dc0bd1ed98bf007c6